icu4c/source/test/intltest/ucdtest.cpp

```diff
@@ -1024,3 +1027,95 @@ void UnicodeTest::TestIDCompatMath() {
+        return u_getPropertyValueEnum((UProperty) property, name);
+    }
+};
+
+U_NAMESPACE_END
+
+void UnicodeTest::TestBinaryPropertyUsingPpucd() {
+    IcuTestErrorCode errorCode(*this, "TestBinaryPropertyUsingPpucd()");
+
+    // Initialize PPUCD parsing object using file in repo and using
+    // property names present in built-in data in ICU
+    char buffer[500];
+    // get path to `source/data/unidata/` including trailing `/`
+    char *unidataPath = getUnidataPath(buffer);
+    if(unidataPath == nullptr) {
+        errln("exiting early because unable to open ppucd.txt from ICU source tree");
+        return;
+    }
+    CharString ppucdPath(unidataPath, errorCode);
+    ppucdPath.appendPathPart("ppucd.txt", errorCode);    
+    PreparsedUCD ppucd(ppucdPath.data(), errorCode);
+    if(errorCode.isFailure()) {
+        errln("unable to open %s - %s\n",
+            ppucdPath.data(), errorCode.errorName());
+        return;
+    }
+    BuiltInPropertyNames builtInPropNames;
+    ppucd.setPropertyNames(&builtInPropNames);
+
+    // Define which binary properties we want to compare
+    constexpr UProperty propsUnderTest[] = {
+        UCHAR_ID_COMPAT_MATH_START,
+        UCHAR_ID_COMPAT_MATH_CONTINUE,
+    };
+
+    // Allocate & initialize UnicodeSets per binary property from PPUCD data
+    UnicodeSet ppucdPropSets[std::size(propsUnderTest)];
+
+    // Iterate through PPUCD file, accumulating each line's data into each UnicodeSet per property
+    PreparsedUCD::LineType lineType;
+    UnicodeSet newValues;
+    while((lineType=ppucd.readLine(errorCode))!=PreparsedUCD::NO_LINE) {
+        if(ppucd.lineHasPropertyValues()) {
+            const UniProps *lineProps=ppucd.getProps(newValues, errorCode);
+
+            for(uint32_t i = 0; i < std::size(propsUnderTest); i++) {
+                UProperty prop = propsUnderTest[i];
+                if (!newValues.contains(prop)) {
+                    continue;
+                }
+                if (lineProps->binProps[prop]) {
+                    ppucdPropSets[i].add(lineProps->start, lineProps->end);
+                } else {
+                    ppucdPropSets[i].remove(lineProps->start, lineProps->end);
+                }
+            }
+        }
+        if(errorCode.isFailure()) {
+            break;
+        }
+    }
+
+    if(errorCode.isFailure()) {
+        errln("exiting early due to parsing error");
+        return;
+    }
+
+    // Assert that the PPUCD data and the ICU data are equivalent for all properties
+    for(uint32_t i = 0; i < std::size(propsUnderTest); i++) {
+        UnicodeSet icuPropSet;
+        icuPropSet.applyIntPropertyValue(propsUnderTest[i], 1, errorCode);
+        std::string msg = 
+            std::string("")
+            + "ICU & PPUCD versions of property "
+            + u_getPropertyName(propsUnderTest[i], U_LONG_PROPERTY_NAME)
+            + ")";
+        assertTrue(msg.c_str(), ppucdPropSets[i] == icuPropSet);
```

Comment:
optional, except definitely balance parentheses:
```suggestion
        UnicodeSet icuPropSet;
        UProperty prop = propsUnderTest[i];
        icuPropSet.applyIntPropertyValue(prop, 1, errorCode);
        std::string msg = 
            std::string()
            + "ICU & PPUCD versions of property "
            + u_getPropertyName(prop, U_LONG_PROPERTY_NAME);
        assertTrue(msg.c_str(), ppucdPropSets[i] == icuPropSet);
```